Streamice update for PETSc 3.8+

dngoldberg opened the pull request:
Since pre-github the streamice package has contained wrappers to the PETSc (Portable, Extensible Toolkit for Scientific Computation) to make use of preconditioners and linear solvers. In most cases, this gives approximately 3-4x speedup of the velocity solver, and greatly accelerates the ice-sheet adjoint solver by caching preconditioner progress. 

With Petsc 3.8, include files could no longer be used and modules need to be used instead, and there have been some changes to the API. 

Changes have had to be made to all streamice subroutines which directly make calls to the PETSc library:
streamice_cg_solver_petsc.F
streamice_initialise_petsc.F
streamice_finalize_petsc.F
streamice_petscmatdestroy.F

As petsc objects can no longer be stored in common blocks (https://lists.mcs.anl.gov/pipermail/petsc-users/2018-March/034825.html) a module file has been added (streamice_arg_common_petsc.F) to replace STREAMICE_PETSC.h when petsc 3.8 or above is used. The name of the module file is important: in OpenAD compilations there does not seem to be a way to force module files to be compiled before source files in the final compilation of transformed source. The file needs to be compiled before streamice_cg_solver_petsc.F.

Additionally, the runtime parameter `streamice_use_petsc` is set to `false` by default. Additional fixes are made so that PETSc functionality can be used without MPI.

## Does this PR introduce a breaking change? 
It does not, as there are no verifications in which the petsc functionality of PETSc is called.

## Suggested addition to `tag-index`
pkg/streamice: modifications to streamice_cg_solver_petsc.F, streamice_initialise_petsc.F, streamice_finalize_petsc.F, streamice_petscmatdestroy.F, streamice_readparms.F streamice_petsc_numerate.F and addition of module file streamice_arg_common_petsc.F to allow use with PETSc 3.8 and above, and in addition to ensure that PETSc is not used at runtime by default, and that PETSc can be used without MPI.

mjlosch commented on pkg/streamice/streamice_arg_common_petsc.F:
```diff
@@ -0,0 +1,42 @@
+#include "petsc/finclude/petscpc.h"
+      use petscpc
+
+      Mat matrix 
```
Is this a fortran statement within a module? I have never seen that.

dngoldberg replied:
@mjlosch do refer to line 19 or 20? 

this material used to appear in a header file -- allowing the (incomplete or full) krylov solver objects to appear in common blocks and thus cache the (incomplete or full) L-U factorisation of the matrix. petsc objects can no longer appear in common blocks (https://lists.mcs.anl.gov/pipermail/petsc-users/2018-March/034825.html) so i found a module to be the only solution. it seems to compile with a range of compilers..

dngoldberg replied:
This is a declaration -- Mat is a fortran derived type.

mjlosch commented on pkg/streamice/streamice_cg_solve_petsc.F:
```diff
@@ -23,23 +23,38 @@ SUBROUTINE STREAMICE_CG_SOLVE_PETSC(
+
+
+#ifdef ALLOW_PETSC
+#ifdef STREAMICE_PETSC_3_8
```
it helps me to have the ifdefs indented like this:
```suggestion
# ifdef STREAMICE_PETSC_3_8
```

dngoldberg replied:
@mjlosch im not sure if this means that all ifdefs, or possibly all cpp directives, should have a single space. before making these quite sweeping changes, would it be possible to see a style guide in case i am misunderstanding?

mjlosch commented on pkg/streamice/streamice_cg_solve_petsc.F:
```diff
@@ -23,23 +23,38 @@ SUBROUTINE STREAMICE_CG_SOLVE_PETSC(
+
+
+#ifdef ALLOW_PETSC
+#ifdef STREAMICE_PETSC_3_8
```
it helps me to have the ifdefs indented like this:
```suggestion
# ifdef STREAMICE_PETSC_3_8
```

dngoldberg replied:
@mjlosch im not sure if this means that all ifdefs, or possibly all cpp directives, should have a single space. before making these quite sweeping changes, would it be possible to see a style guide in case i am misunderstanding?

dngoldberg replied:
This is a declaration -- Mat is a fortran derived type.

mjlosch commented on pkg/streamice/streamice_cg_solve_petsc.F:
```diff
@@ -353,19 +395,17 @@ SUBROUTINE STREAMICE_CG_SOLVE_PETSC(
        call PCSetType(pc,PCLU,ierr)
-       call PCFactorSetMatSolverPackage(pc,MATSOLVERMUMPS,ierr)
-       call PCFactorSetUpMatSolverPackage(pc,ierr)
+       call PCFactorSetMatSolverType(pc,MATSOLVERMUMPS,ierr)
```
is this change related to pets3.8?

mjlosch commented on pkg/streamice/streamice_cg_solve_petsc.F:
```diff
@@ -353,19 +395,17 @@ SUBROUTINE STREAMICE_CG_SOLVE_PETSC(
        call PCSetType(pc,PCLU,ierr)
-       call PCFactorSetMatSolverPackage(pc,MATSOLVERMUMPS,ierr)
-       call PCFactorSetUpMatSolverPackage(pc,ierr)
+       call PCFactorSetMatSolverType(pc,MATSOLVERMUMPS,ierr)
```
is this change related to pets3.8?

dngoldberg replied:
@mjlosch  yes, i was able to determine this was the appropriate change by googling.

mjlosch commented on pkg/streamice/streamice_finalize_petsc.F:
```diff
@@ -10,17 +10,35 @@ SUBROUTINE STREAMICE_FINALIZE_PETSC
+#include "STREAMICE_PETSC.h"
+#endif
+#endif
+      IMPLICIT NONE
```
The implicit none in F77 should come before including anything. This was done so in [streamice_cg_solve_petsc.F](https://github.com/MITgcm/MITgcm/pull/630/files#diff-a3a75347ba43cd790ff6d10183055ca598aa246dd0104cbcec81a24a1df35ef3), why not here?

dngoldberg replied:
streamice_finalize_petsc.F (as well as streamice_petscmatdestroy.F) have been changed to be consistent with streamice_cg_solve_petsc.F wrt placement of IMPLICIT NONE

mjlosch commented on pkg/streamice/streamice_finalize_petsc.F:
```diff
@@ -10,17 +10,35 @@ SUBROUTINE STREAMICE_FINALIZE_PETSC
+#include "STREAMICE_PETSC.h"
+#endif
+#endif
+      IMPLICIT NONE
```
The implicit none in F77 should come before including anything. This was done so in [streamice_cg_solve_petsc.F](https://github.com/MITgcm/MITgcm/pull/630/files#diff-a3a75347ba43cd790ff6d10183055ca598aa246dd0104cbcec81a24a1df35ef3), why not here?

dngoldberg replied:
@mjlosch it also must come before any USE statements, which i hadn't realised before. i have changed to mirror the S/R you mentioned.

mjlosch commented on pkg/streamice/streamice_petscmatdestroy.F:
```diff
@@ -12,20 +12,32 @@ SUBROUTINE STREAMICE_PETSCMATDESTROY(
+#include "petsc/finclude/petscpc.h"
+      use petscpc
+      use STREAMICE_PETSC_mod
+# else
```
`IMPLICIT NONE` is gone. Not needed?

dngoldberg replied:
@mjlosch i think technically not as all variables have been declared but i have changed the code to mirror [this S/R](https://github.com/MITgcm/MITgcm/pull/630/files#diff-a3a75347ba43cd790ff6d10183055ca598aa246dd0104cbcec81a24a1df35ef3)

mjlosch commented on pkg/streamice/streamice_petscmatdestroy.F:
```diff
@@ -12,20 +12,32 @@ SUBROUTINE STREAMICE_PETSCMATDESTROY(
+#include "petsc/finclude/petscpc.h"
+      use petscpc
+      use STREAMICE_PETSC_mod
+# else
```
`IMPLICIT NONE` is gone. Not needed?

mjlosch reviewed (commented):
I cannot run this because I don't have petsc installed, but I went through the code change and have a few comments.

jahn commented:
I've tested this on a version of halfpipe_streamice with the following options set:

```
#define ALLOW_PETSC
#define STREAMICE_PETSC_3_8
```

The version without MPI failed because mpimywid is used in streamice_petsc_numerate but not defined without MPI.  With that fixed, I get 3 or 4 digits (with/without MPI) of agreement for StrmIc compared to the reference output without PETsc.  Not sure whether that is expected.

dngoldberg commented:
Hi @jahn thanks for this. A little surprised by the low accuracy, wonder if compiler settings are affected? But maybe it is expected when linking with add’l libraries. I will try this out for myself..

dngoldberg commented:
hi @jahn a source of disagreement may be because the R/T param `streamice_use_petsc` is `.true.` by default. 

at the moment though im realising that i've never used petsc/streamice without mpi, and doing so is throwing up issues that you have likely discovered already. Im working through these, which might result in modifications to the P/R, but will hopefully enable petsc functionality without MPI, and also hopefully i can verify the experiment with the the CPP's allowing petsc, but with `streamice_use_petsc` set to false. Ill then address martin's comments above.

dngoldberg commented:
hi @mjlosch, @jahn has helped to develop a singularity solution to get around needing to compile petsc locally.

dngoldberg commented:
@jahn @jrscott i have pushed changes that allow compilation with petsc but without mpi. i now run testreport with the following options set:
`#define ALLOW_PETSC`
`#define STREAMICE_PETSC_3_8`
but with `streamice_use_petsc` set to `false` (which is now the default) and without MPI, and get 6 digits of agreement. However, when i run (using the container) with 
`#undef ALLOW_PETSC`
`#undef STREAMICE_PETSC_3_8`
i also get 6 digits of accuracy as well. 

could it be the use of the compilers within the singularity container leading to the difference? and if so, is this actually an issue, or can be accepted that the container is not meant to reproduce the verification, but on the other hand proves that the code can be compiled with petsc?

jahn commented:
I tried different compilers on the master branch: with gfortran 7.5.0 on Ubuntu 18.04 I get 16 digits, with gfortran 9.4.0 on Ubuntu 20.04 I get only 6.  Same with the pull request, but with using PETsc.  So it is definitely the compiler version.

dngoldberg commented:
many thanks for this @jahn. @jrscott, @mjlosch I have updated the P/R introduction to reflect additional changes, and addressed all of Martin's comments (except about style of cpp directives). not sure if anything else is needed?

dngoldberg commented on pkg/streamice/streamice_arg_common_petsc.F:
```diff
@@ -0,0 +1,42 @@
+#include "petsc/finclude/petscpc.h"
+      use petscpc
+
+      Mat matrix 
```
This is a declaration -- Mat is a fortran derived type.

jm-c commented:
I took a (very) quick look and LGTM. Going to merge this PR soon (tomorrow ?) unless someone wants to add something.
